### PR TITLE
Configure CI

### LINKED ISSUE #4: Configure CI
- Setup CI script
- Connect to remote device lab
- Add keystore and secret management
- Add code coverage and reports
- Export checkstyle report
- Export lint report

### PR DESCRIPTION
**Pull request by Numan1617**
Closes #4.

Adds base CI integration including Firebase Cloud Test Lab

### CODE REVIEW

**Review by dmallcott-ustwo — APPROVED** [review]

**Review by mshipton — APPROVED** [review]
Looking great 👍

File: circle.yml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+dependencies:
+  pre:
+    - openssl aes-256-cbc -d -in ci-keys.cipher -k $CI_SECRETS_KEY >> ~/.circlerc
+    - openssl aes-256-cbc -d -in release.keystore.cipher -out release.keystore -k $CI_SECRETS_KEY
+    - if [ ! -e $ANDROID_HOME/build-tools/$ANDROID_BUILD_TOOLS ]; then echo y | android update sdk --no-ui --all --filter build-tools-$ANDROID_BUILD_TOOLS; fi
+    - if [ ! -e $ANDROID_HOME/platforms/android-$ANDROID_PLATFORM ]; then echo y | android update sdk --no-ui --all --filter "android-${ANDROID_PLATFORM}"; fi
+
+  cache_directories:
+    - /usr/local/android-sdk-linux/build-tools/24.0.1
```
Comment (SachseInTheHub) [inline circle.yml:15]: @Numan1617 caching be caching!??!
😸

File: scripts/ci.sh [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+if [ ! -z "$1" ]
+then
+    report_location=$1
+else
+    report_location=`pwd`
+fi
+
+./gradlew --no-daemon --info clean
+./gradlew --no-daemon --info build countDebugDexMethods -PdisablePreDex
+if [ -n "${GCLOUD_SERVICE_KEY+1}" ]; then
```
Comment (SachseInTheHub) [inline scripts/ci.sh:15]: @Numan1617 could you pls add some comments here to explain a bit more off this script!?
kthxbye

File: scripts/remoteTesting.sh [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+    $GCLOUD auth activate-service-account $GCLOUD_SERVICE_ACCOUNT --key-file $HOME/client-secret.json
+    sudo $GCLOUD --quiet components update
+    sudo $GCLOUD --quiet components install beta
+
+    GCLOUD_OUTPUT_LOG=gcloudLog.txt
+
+    $GCLOUD beta test android run --type instrumentation --uri --app app/build/outputs/apk/app-debug.apk --test app/build/outputs/apk/app-debug-androidTest.apk --orientations portrait --project $GCLOUD_PROJECT_ID 2>&1 | tee $GCLOUD_OUTPUT_LOG ; ( exit ${PIPESTATUS[0]} )
+
+    BUCKET_ID=`cat $GCLOUD_OUTPUT_LOG | sed -n -E 's#^.+test-lab-(.+)/.+#\1#p'`
```
Comment (SachseInTheHub) [inline scripts/remoteTesting.sh:24]: @Numan1617 very neat but if you can comment a bit

**Review by SachseInTheHub — COMMENTED** [review]
this is looking pretty epic @Numan1617 !

File: scripts/ci.sh (hunk 2 above)
Comment (mshipton) [reply scripts/ci.sh:15]: Ah, good call - @SachseInTheHub did a nice job of this in another project - perhaps you can share?

Comment (SachseInTheHub) [reply scripts/ci.sh:15]: @mshipton yep, I'll slack it @Numan1617

File: circle.yml (hunk 1 above)
Comment (Numan1617) [reply circle.yml:15]: @SachseInTheHub Qué?

Comment (SachseInTheHub) [reply circle.yml:15]: @Numan1617 is the SDK being cached? I remember we had that we had different opinions about it. Does it actually caches it ? and does that reflect on build times?

File: README.md [hunk 4]
```diff
@@ -15,6 +15,43 @@ This project consists of the following modules:
 * [Android Studio](https://developer.android.com/studio/index.html), Provides the fastest tools for building apps on every type of Android device.
 * [Gradle](https://gradle.org/), An open source build automation system.
 
+## Testing and CI
+
+### Circle
+By default this boilerplate comes configured to run on [CircleCI](circleci.com). The YML file `circle.yml` will setup the dependencies, run the tests configured in `scripts/ci.sh` and export relevant reports and artefacts.
```
Comment (arnau) [inline README.md]: I think you need to put `https://circleci.com` instead of just `circleci.com`

File: README.md [hunk 5]
```diff
@@ -15,6 +15,43 @@ This project consists of the following modules:
...
+
+If you have configured the environment variables defined in the "Firebase cloud testing" the tests will be run on the [Firebase test lab](https://firebase.google.com/docs/test-lab/) otherwise your tests will be run on an emulator.
+
+### Firebase cloud testing
+This boilerplate ships with Firebase cloud testing built in. You simply need to enter the following environment variables to make it work:
+
+- `GCLOUD_SERVICE_ACCOUNT` - The service account email address is listed in the Firebase console here: https://console.firebase.google.com/project/<projectId>/settings/serviceaccounts/adminsdk
+- `GCLOUD_SERVICE_KEY` - This is an export of the contents of the JSON secret file that can be obtained by generating a new private key here: https://console.firebase.google.com/project/<projectId>/settings/serviceaccounts/adminsdk
```
Comment (arnau) [inline README.md]: Add backticks surrounding `https://console.firebase.google.com/project/<projectId>/settings/serviceaccounts/adminsdk` to let markdown parse the placeholder correctly.

File: circle.yml (hunk 1 above)
Comment (Numan1617) [reply circle.yml:15]: @SachseInTheHub There was a lot of back and forth with support@circleci.com. Basically `cache_directories` doesn't expand variables hence my previous solution wasn't working

File: README.md [hunk 6]
```diff
@@ -15,6 +15,43 @@ This project consists of the following modules:
...
+- `GCLOUD_PROJECT_ID` - This is a unique identifier for your Firebase project which can be found in this project's URL: https://console.firebase.google.com/project/<projectId>
```
Comment (arnau) [inline README.md]: Same, backticks in the URL template.

**Review by arnau — COMMENTED** [review]
Great work!  Do you plan to add a note on Blackbox?

File: scripts/ci.sh [hunk 7]
```diff
@@ -1,6 +1,7 @@
 #!/bin/bash
 set -xe
 
+# Get the location for report directory from 1st param or use current directory
 if [ ! -z "$1" ]
 then
```
Comment (arnau) [inline scripts/ci.sh]: Perhaps `then` should be in the same line as the `if` statement to keep consistency with the rest? (or all in two lines)

File: scripts/ci.sh [hunk 8]
```diff
@@ -1,6 +1,7 @@
 #!/bin/bash
```
Comment (arnau) [inline scripts/ci.sh:1]: Probably irrelevant in this case, but I tend to use `#!/usr/bin/env bash` instead (the gradle shellscript uses it too).

File: circle.yml [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+test:
+  pre:
+    - if [ ! -n "${GCLOUD_SERVICE_KEY+1}" ]; then emulator -avd circleci-android24 -no-audio -no-window; fi:
+                background: true
+                parallel: true
+    - if [ ! -n "${GCLOUD_SERVICE_KEY+1}" ]; then circle-android wait-for-boot; fi
+
+  override:
+    - scripts/ci.sh $CIRCLE_TEST_REPORTS
```
Comment (con91) [inline circle.yml:27]: maybe a comment about what ci.sh is doing so you dont have to go look at both? 1st world problems 🌎

**Review by arnau — APPROVED** [review]

File: scripts/ci.sh [hunk 10]
```diff
@@ -2,8 +2,7 @@
 set -xe
 
 # Get the location for report directory from 1st param or use current directory
-if [ ! -z "$1" ]
-then
+if [ ! -z "$1" ] then
```
Comment (arnau) [inline scripts/ci.sh]: Sorry, missing `;`. Should be `if [ ! -z "$1" ]; then`

**Pull request merged**